crates/sui-storage/src/indexes.rs

```diff
@@ -143,64 +151,140 @@ impl IndexStore {
         object_id: KeyT,
-    ) -> SuiResult<Vec<(TxSequenceNumber, TransactionDigest)>> {
-        Ok(index
-            .iter()
-            .skip_to(&(object_id.clone(), TxSequenceNumber::MIN))?
-            .take_while(|((id, _), _)| *id == object_id)
-            .map(|((_, seq), digest)| (seq, digest))
-            .collect())
+        cursor: TxSequenceNumber,
+        limit: Option<usize>,
+        reverse: bool,
+    ) -> SuiResult<Vec<TransactionDigest>> {
+        Ok(if reverse {
+            let iter = index
+                .iter()
+                .skip_prior_to(&(object_id.clone(), cursor))?
+                .reverse()
+                .take_while(|((id, _), _)| *id == object_id)
+                .map(|(_, digest)| digest);
+            if let Some(limit) = limit {
+                iter.take(limit).collect()
+            } else {
+                iter.collect()
+            }
+        } else {
+            let iter = index
+                .iter()
+                .skip_to(&(object_id.clone(), cursor))?
+                .take_while(|((id, _), _)| *id == object_id)
+                .map(|(_, digest)| digest);
+            if let Some(limit) = limit {
+                iter.take(limit).collect()
+            } else {
+                iter.collect()
+            }
+        })
     }
 
     pub fn get_transactions_by_input_object(
         &self,
         input_object: ObjectID,
-    ) -> SuiResult<Vec<(TxSequenceNumber, TransactionDigest)>> {
-        Self::get_transactions_by_object(&self.transactions_by_input_object_id, input_object)
+        cursor: TxSequenceNumber,
+        limit: Option<usize>,
+        reverse: bool,
+    ) -> SuiResult<Vec<TransactionDigest>> {
+        Self::get_transactions_by_object(
+            &self.transactions_by_input_object_id,
+            input_object,
+            cursor,
+            limit,
+            reverse,
+        )
     }
 
     pub fn get_transactions_by_mutated_object(
         &self,
         mutated_object: ObjectID,
-    ) -> SuiResult<Vec<(TxSequenceNumber, TransactionDigest)>> {
-        Self::get_transactions_by_object(&self.transactions_by_mutated_object_id, mutated_object)
+        cursor: TxSequenceNumber,
+        limit: Option<usize>,
+        reverse: bool,
+    ) -> SuiResult<Vec<TransactionDigest>> {
+        Self::get_transactions_by_object(
+            &self.transactions_by_mutated_object_id,
+            mutated_object,
+            cursor,
+            limit,
+            reverse,
+        )
     }
 
     pub fn get_transactions_from_addr(
         &self,
         addr: SuiAddress,
-    ) -> SuiResult<Vec<(TxSequenceNumber, TransactionDigest)>> {
-        Self::get_transactions_by_object(&self.transactions_from_addr, addr)
+        cursor: TxSequenceNumber,
+        limit: Option<usize>,
+        reverse: bool,
+    ) -> SuiResult<Vec<TransactionDigest>> {
+        Self::get_transactions_by_object(&self.transactions_from_addr, addr, cursor, limit, reverse)
```

Comment:
should we rename to `get_transactions_by_object_or_address`

Reply:
this method is more generic, I will rename to `get_transactions_from_index`